data/kernels/diffuse.cl

```diff
@@ -254,34 +258,34 @@ diffuse_pde(read_only image2d_t HF, read_only image2d_t LF,
     find_gradient(neighbour_pixel_LF, gradient);
     find_gradient(neighbour_pixel_HF, laplacian);
 
-    const float4 magnitude_grad = hypot(gradient[0], gradient[1]);
-    const float4 magnitude_lapl = hypot(laplacian[0], laplacian[1]);
-
-    const float4 theta_grad = atan2(gradient[1], gradient[0]);
-    const float4 theta_lapl = atan2(laplacian[1], laplacian[0]);
-
-    const float4 cos_theta_grad = native_cos(theta_grad);
-    const float4 cos_theta_lapl = native_cos(theta_lapl);
-    const float4 sin_theta_grad = native_sin(theta_grad);
-    const float4 sin_theta_lapl = native_sin(theta_lapl);
-
-    const float4 cos_theta_grad_sq = sqf(cos_theta_grad);
-    const float4 sin_theta_grad_sq = sqf(sin_theta_grad);
-    const float4 cos_theta_lapl_sq = sqf(cos_theta_lapl);
-    const float4 sin_theta_lapl_sq = sqf(sin_theta_lapl);
+    const float4 magnitude_grad = native_sqrt(sqf(gradient[0]) + sqf(gradient[1]));
+    const float4 norm_grad = 1.f / fmax(magnitude_grad, NORM_MIN);
```

Comment:
This one took me a while to figure out. 

What you compute here is not the norm, but the hypotenuse, and you divide `grad_x` and `grad_y` by the hypotenuse to get the cos and sin of `arg(grad)`, not to normalize it. Although arithmetically equivalent, the concept is different and it's not immediately clear that `gradient[0]` becomes `cos(arg(gradient))` at the next step. 

I'm not nitpicking, there are sisters methods using Total Variation regularization that also rely on normalizing the laplacian by the magnitude of the gradient, and that look just like that. The difference is here, the magnitude of the gradient is used to compute the angle of the rotation matrix, not to regularize the laplacian, so let's make it clear.

I think when `grad = (0, 0)` we should just force `arg(grad) = 0` by convention and therefore, `cos(0) = 1` and `sin(0) = 1`. In that case, `c2 == 1` exactly. Thresholding by NORM_MIN is inaccurate in that regard.

I would rewrite it as follow (including comments):

```C
const float4 magnitude_grad = hypotf(gradient[0],  gradient[1]);

// Compute cos(arg(grad)) = x / hypot - force arg(grad) = 0 if hypot == 0
gradient[0] = (magnitude_grad != 0.f) ? gradient[0] / magnitude_grad
                                                              : 1.f; // cos(0)
// Compute sin (arg(grad))= y / hypot - force arg(grad) = 0 if hypot == 0
gradient[1] = (magnitude_grad != 0.f) ? gradient[1] / magnitude_grad
                                                              : 0.f; // sin(0)

// Warning : now gradient[2] = { cos(arg(grad)) , sin(arg(arg)) }
```